src/devtools/store.js

```diff
@@ -420,6 +422,28 @@ export default class Store extends EventEmitter {
     this.emit('isProfiling');
   }
 
+  toggleIsCollapsed(id: number, isCollapsed: boolean): void {
+    const element = this.getElementByID(id);
+    if (element !== null) {
+      element.isCollapsed = isCollapsed;
+
+      const weightDelta = isCollapsed ? 1 - element.weight : element.weight - 1;
```

Comment:
This might be more obviously correct:

```js
const oldWeight = element.isCollapsed ? 1 : element.weight;
element.isCollapsed = isCollapsed;
const newWeight = element.isCollapsed ? 1 : element.weight;
const weightDelta = newWeight - oldWeight;
```

Reply:
Thinking more on this, I don't love that by doing delta updates, this allows for errors to compound if any of the logic is wrong.

I think my preference would be to re-derive it each time. One easy way to do this is to have it be a lazy getters (or equivalent):

```js
function getVisibleWeight(el) {
  if (el.cachedVisibleWeight) {
    return el.cachedVisibleWeight;
  }
  return el.cachedVisibleWeight = el.isCollapsed ? 1 : 1 + sum(getVisibleWeight() of children);
}
```

When you change the children of any node or expand/collapse, you invalidate the cached weights of all the ancestors. Notably:
1. You do *not* need to recompute them all eagerly, and it is better not to. That way, if you change two parts of a tree, you only recompute the ancestor once.
2. When you are traversing ancestors, you can stop once you hit an ancestor that is already invalidated. That means you don't need to traverse all the way to the root for every change in a deep tree, only the first.

Reply:
> Thinking more on this, I don't love that by doing delta updates, this allows for errors to compound if any of the logic is wrong.

AFK and so I'm not really responding to the review as a whole, just cherry picking this specific detail of it. In general, if any of the "logic is wrong", the tree is screwed. Everything is tuned for optimal performance (in some cases at the slight cost of readability) as best I know how.

I plan to start adding tests soon to cover the various types of tree mutations to reduce the likelihood of a regression

Reply:
> Everything is tuned for optimal performance (in some cases at the slight cost of readability) as best I know how.

Well, you're using React so I assume you're taking at least _some_ preference for readability/maintainability/verifiable-correctness over the slightly faster possible speed that using solely imperative DOM mutations would provide.

I'm sympathetic to wanting it to be fast which is why I mentioned perf, but when it is possible to write code in a simpler way that is easier to verify the correctness of, I think that's better unless it is an extreme perf hotspot (which I really wouldn't expect for this case).

Reply:
Hmm. We may have to agree to disagree on the right trade off between readability and performance in this case. In the case of the `Store` and op-codes, I intentionally chose a _less readable_ approach so that I could optimize perf as much as possible. (We make a lot of similar trade offs in the React source code.)

Amortizing the cost of these tree mutations is a core part of the design, so that interactions with the tree (especially scrolling and windowing related logic) can be as fast as possible.

Reply:
My suggestion also amortizes the calculations. :)

Fine to stick with your way, but I stand by my suggestion in the first comment then.

Reply:
Thats fair 👍 I don't mind making that change.